### PR TITLE
Optimize LayerNorm Backward dx

### PR DESCRIPTION
**Pull request by guo-ran**
pr2: 优化后向Grad实现：根据cols大小采用分段函数实现以达到最优性能，替代现有的调用cudnn的方法。

### CODE REVIEW

File: oneflow/core/cuda/layer_norm.cuh [hunk 1]
```diff
@@ -815,6 +852,556 @@ inline cudaError_t DispatchLayerNorm(cudaStream_t stream, LOAD load, STORE store
...
+  for (int64_t row = global_thread_group_id * rows_per_access; row < rows;
+       row += step) {
+    ComputeType sum_loss1[rows_per_access];
+    ComputeType sum_loss2[rows_per_access];
+#pragma unroll
+    for (int row_id = 0; row_id < rows_per_access; ++row_id) {
+      const int global_row_id = row + row_id;
+      mean_buf[row_id] = mean[global_row_id];
+      inv_variance_buf[row_id] = inv_variance[global_row_id];
+      sum_loss1[row_id] = 0;
```
Comment (MARD1NO) [inline oneflow/core/cuda/layer_norm.cuh:882]: sum_loss1/2 这两个名字可以想想，或者加点数学公式的注释

File: oneflow/core/cuda/layer_norm.cuh [hunk 2]
```diff
@@ -815,6 +852,556 @@ inline cudaError_t DispatchLayerNorm(cudaStream_t stream, LOAD load, STORE store
...
+      const ComputeType inv_variance_over_cols =
+          inv_variance_buf[row_id] / static_cast<ComputeType>(cols);
+#pragma unroll
+      for (int pack_id = 0; pack_id < pack_per_thread; ++pack_id) {
+        const int col = (pack_id * thread_group_width + lane_id) * pack_size;
+        if (!padding || col < cols) {
+          for (int i = 0; i < pack_size; ++i) {
+            const int col_id = pack_id * pack_size + i;
+            row_dy_buf[col_id] = (cols * row_dy_buf[col_id] - warp_sum_loss1[row_id]
+                                  - (row_x_buf[col_id] - mean_buf[row_id])
```
Comment (MARD1NO) [inline oneflow/core/cuda/layer_norm.cuh:926]: `row_x_buf[col_id] - mean_buf[row_id]` 能和897行的那句话，提取出公共表达式吗

File: oneflow/core/cuda/layer_norm.cuh [hunk 3]
```diff
@@ -815,6 +852,556 @@ inline cudaError_t DispatchLayerNorm(cudaStream_t stream, LOAD load, STORE store
...
+      ComputeType* row_dy_buf = dy_buf[row_id];
+      const ComputeType inv_variance_over_cols =
+          inv_variance_buf[row_id] / static_cast<ComputeType>(cols);
+#pragma unroll
+      for (int pack_id = 0; pack_id < pack_per_thread; ++pack_id) {
+        const int col = (pack_id * thread_group_width + lane_id) * pack_size;
+        if (!padding || col < cols) {
+          for (int i = 0; i < pack_size; ++i) {
+            const int col_id = pack_id * pack_size + i;
+            row_dy_buf[col_id] = (cols * row_dy_buf[col_id] - warp_sum_loss1[row_id]
```
Comment (MARD1NO) [inline oneflow/core/cuda/layer_norm.cuh:925]: `cols * row_dy_buf[col_id]` 为什么要乘一个cols？

**Review by MARD1NO — APPROVED** [review]

**Review by liujuncheng — APPROVED** [review]

**Comment by ouyangyu** [comment]
- ### BERT正确性验证

- https://oneflow-public.oss-cn-beijing.aliyuncs.com/oneflow_test_log/dev_light_actor/bert_v0.4.0_adam_1n8g_sq128_nhl12_nah12_bsz48_iter1.log
- https://oneflow-test.oss-cn-beijing.aliyuncs.com/OneBrain/commit/dev_layer_norm_backward_dx/20211209/ab20740/BERT-lazy/train/1n8g/bert_1n8g_sq128_nhl12_nah12_bsz48_adam_dev_layer_norm_backward_dx_20211209_153632466545891.log


截屏2021-12-09 17 00 17

**Pull request merged**